### PR TITLE
arch-vega,gpu-compute,mem-ruby: SQC Invalidation Support

### PR DESCRIPTION
**Pull request by v-ramadas**
This PR adds support for SQC (GPU I-cache) invalidation to the GPU model. It does this by updating the GPU-VIPER-SQC protocol to support flushes, the sequencer model to send out invalidates and the gpu compute model to send invalidates and handle responses. It also adds support for S_ICACHE_INV, a VEGA ISA instruction that invalidates the entire GPU I-cache. Additionally, the PR modifies the kernel start behavior to invalidate the I-cache too. It previously invalidated only the L1 D-cache.

### CODE REVIEW

File: src/arch/amdgpu/vega/insts/instructions.cc [hunk 1]
```diff
@@ -4818,6 +4818,10 @@ namespace VegaISA
     Inst_SOPP__S_ICACHE_INV::Inst_SOPP__S_ICACHE_INV(InFmt_SOPP *iFmt)
         : Inst_SOPP(iFmt, "s_icache_inv")
     {
+        setFlag(MemBarrier);
+        setFlag(GPUStaticInst::MemSync);
+        setFlag(MemSync);
+
```
Comment (mattsinc) [inline src/arch/amdgpu/vega/insts/instructions.cc]: extra line of space, remove

File: src/gpu-compute/fetch_unit.cc [hunk 2]
```diff
@@ -388,6 +388,29 @@ FetchUnit::FetchBufDesc::flushBuf()
...
+    readPtr = bufStart;
+
+    for (int i = 0; i < fetchDepth; ++i) {
+        freeList.push_back(bufStart + i * cacheLineSize);
+    }
+
+    DPRINTF(GPUFetch, "WF[%d][%d]: Id%d Fetch dropped, flushing fetch "
+            "buffer\n", wavefront->simdId, wavefront->wfSlotId,
+            wavefront->wfDynId);
+
```
Comment (mattsinc) [inline src/gpu-compute/fetch_unit.cc]: extra line of space, remove

File: src/gpu-compute/fetch_unit.cc [hunk 3]
```diff
@@ -471,6 +494,13 @@ FetchUnit::FetchBufDesc::reserveBuf(Addr vaddr)
 void
 FetchUnit::FetchBufDesc::fetchDone(Addr vaddr)
 {
+    if (vaddr == 0) {
```
Comment (mattsinc) [inline src/gpu-compute/fetch_unit.cc]: you should probably add a comment here explaining what vaddr == 0 represents.

A bigger debate is if we can really guarantee vaddr == 0 only for s_icache_inv, but for now probably ok.

File: src/gpu-compute/scalar_memory_pipeline.cc [hunk 4]
```diff
@@ -160,4 +160,55 @@ ScalarMemPipeline::issueRequest(GPUDynInstPtr gpuDynInst)
...
+                0, 0, 0, computeUnit.requestorId(), 0, gpuDynInst->wfDynId);
+    } else {
+        req->requestorId(computeUnit.requestorId());
+    }
+
+    req->setPaddr(0);
+
+    PacketPtr pkt = nullptr;
+
+    if (kernelMemSync) {
```
Comment (mattsinc) [inline src/gpu-compute/scalar_memory_pipeline.cc:188]: add comment here explaining if/else

File: src/gpu-compute/scalar_memory_pipeline.cc [hunk 5]
```diff
@@ -160,4 +160,55 @@ ScalarMemPipeline::issueRequest(GPUDynInstPtr gpuDynInst)
...
+    assert(gpuDynInst->isScalar());
+
+    if (!req) {
+        req = std::make_shared<Request>(
+                0, 0, 0, computeUnit.requestorId(), 0, gpuDynInst->wfDynId);
+    } else {
+        req->requestorId(computeUnit.requestorId());
+    }
+
+    req->setPaddr(0);
```
Comment (mattsinc) [inline src/gpu-compute/scalar_memory_pipeline.cc:181]: add comment explaining this

File: src/gpu-compute/scalar_memory_pipeline.cc [hunk 6]
```diff
@@ -160,4 +160,55 @@ ScalarMemPipeline::issueRequest(GPUDynInstPtr gpuDynInst)
...
+        pkt = new Packet(req, MemCmd::MemSyncReq);
+        pkt->pushSenderState(
+                new ComputeUnit::SQCPort::SenderState(
+                    gpuDynInst->wavefront(), nullptr));
+
+        ComputeUnit::SQCPort::MemReqEvent *sqc_event =
+                new ComputeUnit::SQCPort::MemReqEvent
+                (computeUnit.sqcPort, pkt);
+
+        computeUnit.schedule(
```
Comment (mattsinc) [inline src/gpu-compute/scalar_memory_pipeline.cc]: probably the schedule can be after the if-else, since it's the same in both?  And maybe the sqc_event creation too?

File: src/gpu-compute/shader.cc [hunk 7]
```diff
@@ -221,6 +221,11 @@ Shader::prepareInvalidate(HSAQueueEntry *task) {
         // all necessary INV flags are all set now, call cu to execute
         cuList[i_cu]->doInvalidate(req, task->dispatchId());
 
+
+        if ((i_cu % 4) == 0) {
```
Comment (mattsinc) [inline src/gpu-compute/shader.cc]: Do we have a variable in the Python for this?  4 is a magic number because of how many CUs share an SQC.  To make this code more resilient we should really add/use such a variable to avoid problems later.

Comment (mattsinc) [inline src/gpu-compute/shader.cc]: also add comment here explaining logic

File: src/mem/ruby/protocol/GPU_VIPER-SQC.sm [hunk 8]
```diff
@@ -313,6 +320,11 @@ machine(MachineType:SQC, "GPU SQC (L1 I Cache)")
     APPEND_TRANSITION_COMMENT(cache_entry.DataBlk);
   }
 
+  action(inv_invDone, "inv", desc="local inv done") {
+    sequencer.invL1Callback();
+  }
+
```
Comment (mattsinc) [inline src/mem/ruby/protocol/GPU_VIPER-SQC.sm:326]: extra line of space, remove

File: src/mem/ruby/system/Sequencer.cc [hunk 9]
```diff
@@ -348,6 +350,10 @@ Sequencer::insertRequest(PacketPtr pkt, RubyRequestType primary_type,
         return RequestStatus_Ready;
     }
 
+    if (pkt->cmd == MemCmd::MemSyncReq) {
+        return RequestStatus_Aliased;
```
Comment (mattsinc) [inline src/mem/ruby/system/Sequencer.cc:359]: What is this for?

**Review by mattsinc — CHANGES_REQUESTED** [review]
In general looks good, but some questions/comments/changes.  Nice job!

File: src/arch/amdgpu/vega/insts/instructions.cc (hunk 1 above)
Comment (v-ramadas) [reply src/arch/amdgpu/vega/insts/instructions.cc]: Done

File: src/mem/ruby/protocol/GPU_VIPER-SQC.sm (hunk 8 above)
Comment (v-ramadas) [reply src/mem/ruby/protocol/GPU_VIPER-SQC.sm:326]: Done

File: src/gpu-compute/fetch_unit.cc (hunk 3 above)
Comment (v-ramadas) [reply src/gpu-compute/fetch_unit.cc]: Added comment

File: src/gpu-compute/scalar_memory_pipeline.cc (hunk 5 above)
Comment (v-ramadas) [reply src/gpu-compute/scalar_memory_pipeline.cc:181]: Done

File: src/gpu-compute/scalar_memory_pipeline.cc (hunk 4 above)
Comment (v-ramadas) [reply src/gpu-compute/scalar_memory_pipeline.cc:188]: Done

File: src/gpu-compute/scalar_memory_pipeline.cc (hunk 6 above)
Comment (v-ramadas) [reply src/gpu-compute/scalar_memory_pipeline.cc]: Done

File: src/gpu-compute/shader.cc (hunk 7 above)
Comment (v-ramadas) [reply src/gpu-compute/shader.cc]: We have one in the top level config script but did not in the Shader/Compute Unit models. I added one and use it in this logic instead. It gets its value from the top level config script. I wanted to do this in my initial commit but completely forgot about it. Thanks for pointing it out.

Comment (v-ramadas) [reply src/gpu-compute/shader.cc]: Done

File: src/mem/ruby/system/Sequencer.cc (hunk 9 above)
Comment (v-ramadas) [reply src/mem/ruby/system/Sequencer.cc:359]: Added a comment explaining it

**Comment by v-ramadas** [comment]
@mattsinc, thanks for reviewing! I have made the updates you suggested and added comments accordingly.

File: src/mem/ruby/protocol/MOESI_AMD_Base-CorePair.sm [hunk 10]
```diff
@@ -575,9 +581,12 @@ machine(MachineType:CorePair, "CP-like Core Coherence")
                   if (in_msg.Type == RubyRequestType:LD) {
                     trigger(Event:C1_Load_L1miss, in_msg.LineAddress,
                             cache_entry, tbe);
-                  } else {
+                  } else if ((in_msg.Type == RubyRequestType:ST) || (in_msg.Type == RubyRequestType:ATOMIC)) {
```
Comment (mattsinc) [inline src/mem/ruby/protocol/MOESI_AMD_Base-CorePair.sm]: What does this have to do with the SQC commit?

File: src/mem/ruby/protocol/MOESI_AMD_Base-CorePair.sm [hunk 11]
```diff
@@ -613,9 +625,12 @@ machine(MachineType:CorePair, "CP-like Core Coherence")
                   if (in_msg.Type == RubyRequestType:LD) {
                     trigger(Event:C0_Load_L1miss, in_msg.LineAddress,
                         cache_entry, tbe);
-                  } else {
+                  } else if ((in_msg.Type == RubyRequestType:ST) || (in_msg.Type == RubyRequestType:ATOMIC)) {
```
Comment (mattsinc) [inline src/mem/ruby/protocol/MOESI_AMD_Base-CorePair.sm]: same here

File: src/gpu-compute/shader.cc (hunk 7 above)
Comment (mattsinc) [reply src/gpu-compute/shader.cc]: Probably we'd need to update that script too to reflect this change, but it shouldn't be part of this patch

File: src/gpu-compute/compute_unit.cc [hunk 12]
```diff
@@ -1042,6 +1069,28 @@ ComputeUnit::SQCPort::recvReqRetry()
...
+void
+ComputeUnit::SQCPort::MemReqEvent::process()
+{
+    SenderState *sender_state = safe_cast<SenderState*>(pkt->senderState);
+    [[maybe_unused]] ComputeUnit *compute_unit = sqcPort.computeUnit;
+
+    if (pkt->req->systemReq()) {
+        assert(compute_unit->shader->systemHub);
+        SystemHubEvent *resp_event = new SystemHubEvent(pkt, &sqcPort);
+        compute_unit->shader->systemHub->sendRequest(pkt, resp_event);
```
Comment (abmerop) [inline src/gpu-compute/compute_unit.cc]: I don't think this path is possible.  If it is, since the SystemHubEvent process() does nothing we will have requests silently dropped.  I would instead make this an assert that it is not a system request.  If we see the assert in the future we'll know exactly where to go back and fix it

File: src/gpu-compute/compute_unit.hh [hunk 13]
```diff
@@ -674,6 +675,41 @@ class ComputeUnit : public ClockedObject
...
+            {
+                setFlags(Event::AutoDelete);
+            }
+
+            void
+            process()
+            {
+            }
+        };
+
```
Comment (abmerop) [inline src/gpu-compute/compute_unit.hh]: Probably not needed due to above comment

File: src/gpu-compute/fetch_unit.cc [hunk 14]
```diff
@@ -388,6 +388,28 @@ FetchUnit::FetchBufDesc::flushBuf()
...
+
+    for (int i = 0; i < fetchDepth; ++i) {
+        freeList.push_back(bufStart + i * cacheLineSize);
+    }
+
+    DPRINTF(GPUFetch, "WF[%d][%d]: Id%d Fetch dropped, flushing fetch "
+            "buffer\n", wavefront->simdId, wavefront->wfSlotId,
+            wavefront->wfDynId);
+}
+
```
Comment (abmerop) [inline src/gpu-compute/fetch_unit.cc]: This seems to be exactly the same as flushBuf(), is there any reason not to use that?  I realize the terms have different meanings, but for a read-only buffer they are functionally the same

File: src/gpu-compute/fetch_unit.cc [hunk 15]
```diff
@@ -471,6 +493,18 @@ FetchUnit::FetchBufDesc::reserveBuf(Addr vaddr)
 void
 FetchUnit::FetchBufDesc::fetchDone(Addr vaddr)
 {
+    // If the return vaddr is 0, then it belongs to an SQC invalidation
+    // request. This request calls incLGKMInstsIssued() function in its
+    // execution path. Since there is no valid memory return response
+    // associated with this instruction, decLGKMInstsIssued() is not
+    // executed. Do this here to decrement the counter and invalidate
+    // all buffers
+    if (vaddr == 0) {
```
Comment (abmerop) [inline src/gpu-compute/fetch_unit.cc]: This method seems to be called in one place only which has access to the packet.  I think it would be safer to change the method to pass the packet, then here it could check if it is a MemSyncResp (or whatever SQC invalidate response is)

File: src/gpu-compute/compute_unit.cc (hunk 12 above)
Comment (v-ramadas) [reply src/gpu-compute/compute_unit.cc]: Thanks for letting me know. I was not sure about this part earlier. I have now changed this to an assert and the code still works

File: src/gpu-compute/compute_unit.hh (hunk 13 above)
Comment (v-ramadas) [reply src/gpu-compute/compute_unit.hh]: Removed

File: src/gpu-compute/fetch_unit.cc (hunk 15 above)
Comment (v-ramadas) [reply src/gpu-compute/fetch_unit.cc]: I updated the function to use the packet command instead. Thanks for the suggestion!

File: src/mem/ruby/protocol/MOESI_AMD_Base-CorePair.sm (hunk 10 above)
Comment (v-ramadas) [reply src/mem/ruby/protocol/MOESI_AMD_Base-CorePair.sm]: I incorrectly added this file. These changes were in my stash and I applied it wrongly added it when making review changes. Removed this from the commit now.

These changes are from when I added Flush support for L1,L2. Since GPUFS uses KVM instead of the CorePair, I didn't get to test this. Haven't made a commit for these changes because of this reason

**Review by abmerop — DISMISSED** [review]

File: src/gpu-compute/fetch_unit.cc (hunk 15 above)
Comment (mattsinc) [reply src/gpu-compute/fetch_unit.cc]: good call, @abmerop !

**Review by mattsinc — DISMISSED** [review]

File: src/arch/amdgpu/vega/insts/instructions.cc
Comment (abmerop) [inline src/arch/amdgpu/vega/insts/instructions.cc:1]: Forgot this file no longer exists.  It would now go in the sopp.cc file

Comment (mattsinc) [reply src/arch/amdgpu/vega/insts/instructions.cc:1]: good point @abmerop , @v-ramadas you must not have updated to the head of develop before making your changes?

Comment (v-ramadas) [reply src/arch/amdgpu/vega/insts/instructions.cc:1]: I was at the head of develop when I first made the branch. This was before the commit to break up instructions.cc into multiple files was merged though. I will update the PR with this change now

**Comment by v-ramadas** [comment]
@mattsinc @abmerop, I rebased my branch to the top of develop and wrote the S_ICACHE_INV behavior in sopp.cc. I also removed the older commit that was causing the merge issue. I've pushed the changes to this PR after verifying my test still works.

**Review by abmerop — APPROVED** [review]

**Review by mattsinc — APPROVED** [review]
